pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1266,19 +1282,43 @@ func (vs *VSphere) CreateVolume(volumeOptions *VolumeOptions) (volumePath string
 	dc, err := f.Datacenter(ctx, vs.cfg.Global.Datacenter)
 	f.SetDatacenter(dc)
 
+	if volumeOptions.StoragePolicyName != "" {
+		// Get the pbm client
+		pbmClient, err := pbm.NewClient(ctx, vs.client.Client)
+		if err != nil {
+			return "", err
+		}
+		volumeOptions.StoragePolicyID, err = pbmClient.ProfileIDByName(ctx, volumeOptions.StoragePolicyName)
+		if err != nil {
+			return "", err
+		}
+		// Get the resource pool for current node.
+		resourcePool, err := vs.getCurrentNodeResourcePool(ctx, dc)
+		if err != nil {
+			return "", err
+		}
+
+		dsRefs, err := vs.GetCompatibleDatastores(ctx, pbmClient, resourcePool, volumeOptions.StoragePolicyID)
+		if err != nil {
+			return "", err
+		}
+
+		if volumeOptions.Datastore != "" {
+			if !IsUserSpecifiedDatastoreCompatible(dsRefs, volumeOptions.Datastore) {
+				return "", fmt.Errorf("User specified datastore: %q is not compatible with the StoragePolicy: %q requirements", volumeOptions.Datastore, volumeOptions.StoragePolicyName)
```

Comment:
Don't we want to say why it is not compatible? SPBM gives reasons why a particular datastore is incompatible.